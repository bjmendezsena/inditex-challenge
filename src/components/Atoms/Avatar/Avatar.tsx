/* eslint-disable react-hooks/exhaustive-deps */
import { useMemo } from "react";
import { type avatarAnatomy as ChakraAvatarParts } from "@chakra-ui/anatomy";
import {
  Avatar as ChakraAvatar,
  type AvatarProps as ChakraAvatarProps,
} from "@chakra-ui/avatar";
import {
  forwardRef,
  useMultiStyleConfig as useChakraMultiStyleConfig,
} from "@chakra-ui/system";
import type {
  SystemStyleObject as ChakraSystemStyleObject,
  ComponentWithAs,
} from "@chakra-ui/system";
import { SkeletonCircle as ChakraSkeletonCircle } from "@chakra-ui/react";

import { type BaseMultiStyleConfiguration } from "../../../theme";
import _merge from "lodash/merge";

export const avatarPartsList: typeof ChakraAvatarParts.keys = [
  "container",
  "badge",
  "label",
  "group",
  "excessLabel",
];

type Parts = typeof avatarPartsList;

export interface AvatarThemeConfiguration
  extends BaseMultiStyleConfiguration<Parts> {
  baseStyle: ({
    name,
    badgeColor,
  }: {
    badgeColor?: string;
    name?: string;
  }) => Partial<AvatarThemeStyle>;
}

type AvatarThemeStyle = Record<Parts[number], ChakraSystemStyleObject>;
export interface AvatarProps extends Omit<ChakraAvatarProps, "children"> {
  isLoading?: boolean;
}

export const Avatar: ComponentWithAs<"span", AvatarProps> = forwardRef<
  AvatarProps,
  "span"
>(({ size, name, src, variant, isLoading, sx, ...props }: AvatarProps, ref) => {
  const styles = useChakraMultiStyleConfig("Avatar", {
    name: name && !src,
    size,
    variant,
  }) as AvatarThemeStyle;

  const avatarStyles = useMemo(() => _merge(styles.container, sx),[sx])

  if (isLoading) {
    return (
      <ChakraSkeletonCircle sx={styles.container} size={size} {...props} />
    );
  }


  return (
    <ChakraAvatar
      ref={ref}
      name={name}
      src={src}
      sx={avatarStyles}
      variant={variant}
      {...props}
    />
  );
});
