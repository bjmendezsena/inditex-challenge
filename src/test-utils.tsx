import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === "test" ? () => {} : console.error,
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const DefaultTheme = extendTheme(theme);

export const render = (ui: JSX.Element, theme = DefaultTheme): RenderResult => {
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
