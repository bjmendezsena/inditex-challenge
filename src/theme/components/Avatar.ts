import {
  type AvatarThemeConfiguration,
  avatarPartsList,
} from "../../components";

const baseStyle: AvatarThemeConfiguration["baseStyle"] = ({
  name,
  badgeColor,
}) => ({
  container: {
    bg: "white",
    fontWeight: "medium",
    svg: {
      color: "grey.400",
    },
  },
  badge: {
    border: "none",
    transform: name ? "translate(0%, 0%)" : "translate(25%, 0%)",
    boxShadow: "0 0 0 0.15em var(--chakra-colors-white)",
    bg: badgeColor ? badgeColor : "green.500",
  },
});

const sizes: AvatarThemeConfiguration["sizes"] = {
  md: ({ name }) => ({
    container: {
      width: 8,
      height: 8,
    },
    badge: {
      boxSize: name ? 2 : 3.5,
    },
  }),
  lg: ({ name }) => ({
    container: {
      width: "80px",
      height: "80px",
    },
    badge: {
      boxSize: name ? 2.5 : 4,
    },
  }),
};

export const ConfigAvatar: AvatarThemeConfiguration = {
  parts: avatarPartsList,
  baseStyle,
  sizes,
  variants: {
    light: () => ({
      container: {
        color: "black",
      },
    }),
    dark: () => ({
      container: {
        color: "white",
      },
    }),
  },

  defaultProps: {
    size: "md",
    variant: "light",
  },
};
