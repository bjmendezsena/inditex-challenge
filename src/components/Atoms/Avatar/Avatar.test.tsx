import { render } from "../../../test-utils";
import { Avatar, type AvatarProps } from "./Avatar";

describe(`<${Avatar.name} />`, () => {
  const name = "John Doe";
  const src = "https://placekitten.com/200/200";
  const size2xs = { size: "2xs" };
  const sizeXs = { size: "xs" };
  const sizeSm = { size: "sm" };
  const sizeMd = { size: "md" };
  const sizeLg = { size: "lg" };
  const sizeXl = { size: "xl" };
  const size2Xl = { size: "2xl" };
  const size3Xl = { size: "3xl" };
  const factoryComponent = (props: AvatarProps) =>
    render(<Avatar {...props} />);

  test("renders correctly", () => {
    const view = factoryComponent({});
    expect(view.asFragment()).toMatchSnapshot();
  });

  test("renders correctly with size 2xs", () => {
    const view = factoryComponent(size2xs);
    expect(view.asFragment()).toMatchSnapshot("Using size 2xs");
  });
  test("renders correctly with size xs", () => {
    const view = factoryComponent(sizeXs);
    expect(view.asFragment()).toMatchSnapshot("Using size xs");
  });
  test("renders correctly with size sm", () => {
    const view = factoryComponent(sizeSm);
    expect(view.asFragment()).toMatchSnapshot("Using size sm");
  });
  test("renders correctly with size md", () => {
    const view = factoryComponent(sizeMd);
    expect(view.asFragment()).toMatchSnapshot("Using size md");
  });
  test("renders correctly with size lg", () => {
    const view = factoryComponent(sizeLg);
    expect(view.asFragment()).toMatchSnapshot("Using size lg");
  });
  test("renders correctly with size xl", () => {
    const view = factoryComponent(sizeXl);
    expect(view.asFragment()).toMatchSnapshot("Using size xl");
  });

  test("renders correctly with size 2xl", () => {
    const view = factoryComponent(size2Xl);
    expect(view.asFragment()).toMatchSnapshot("Using size 2xl");
  });

  test("renders correctly with size 3xl", () => {
    const view = factoryComponent(size3Xl);
    expect(view.asFragment()).toMatchSnapshot("Using size 3xl");
  });

  test("renders correctly with initials", () => {
    const view = factoryComponent({ name });
    expect(view.asFragment()).toMatchSnapshot("Using initials");
  });

  test("renders correctly with image", () => {
    const view = factoryComponent({ src });
    expect(view.asFragment()).toMatchSnapshot("Using image");
  });


  test("renders correctly with isLoading", () => {
    const view = factoryComponent({ isLoading: true });
    expect(view.asFragment()).toMatchSnapshot("Using isLoading");
  });
});
