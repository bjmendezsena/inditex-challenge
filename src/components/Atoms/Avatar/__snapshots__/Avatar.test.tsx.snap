// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<undefined /> renders correctly 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-1oojbg6"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with image: Using image 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-1oojbg6"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with initials: Using initials 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-1ufy55n"
  >
    <div
      aria-label="John Doe"
      class="chakra-avatar__initials css-1ebyn6"
      role="img"
    >
      JD
    </div>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with isLoading: Using isLoading 1`] = `
<DocumentFragment>
  <div
    class="chakra-skeleton css-1yzef60"
  />
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size 2xl: Using size 2xl 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-dhov6o"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size 2xs: Using size 2xs 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-lwzsky"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size 3xl: Using size 3xl 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-12zh7x"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size lg: Using size lg 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-12db53i"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size md: Using size md 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-1oojbg6"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size sm: Using size sm 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-1kcjn2d"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size xl: Using size xl 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-19b2sf2"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<undefined /> renders correctly with size xs: Using size xs 1`] = `
<DocumentFragment>
  <span
    class="chakra-avatar css-1a38bvj"
  >
    <svg
      aria-label=" avatar"
      class="chakra-avatar__svg css-16ite8i"
      role="img"
      viewBox="0 0 128 128"
    >
      <path
        d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
        fill="currentColor"
      />
      <path
        d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
        fill="currentColor"
      />
    </svg>
  </span>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
